### PR TITLE
Include override configuration for Prefect Agent

### PR DESCRIPTION
**Pull request by viniciusdc**

### CODE REVIEW

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/kubernetes.tf [hunk 1]
```diff
@@ -235,6 +235,9 @@ module "prefect" {
   {% if cookiecutter.prefect.image is defined -%}
   image                = "{{ cookiecutter.prefect.image }}"
   {% endif -%}
+  {% if cookiecutter.prefect.overrides is defined %}
+  overrides            = ["{{ cookiecutter.prefect.overrides }}"]
```
Comment (costrouc) [inline qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/kubernetes.tf:239]: `"{{ cookiecutter.prefect.overrides | jsonify }}"`

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf [hunk 2]
```diff
@@ -3,6 +3,9 @@ resource "helm_release" "prefect" {
   namespace = var.namespace
   chart     = "${path.module}/chart"
```
Comment (costrouc) [inline qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf:4]: Remove the helm chart. See this forthe chart name and url https://github.com/PrefectHQ/server/tree/master/helm/prefect-server#installing-released-versions.

**Review by costrouc — CHANGES_REQUESTED** [review]
Additionally there need to be docs on how to specify the overrides within the qhub-config.yaml in the configuration docs.

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf (hunk 2 above)
Comment (viniciusdc) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf:4]: Sure, I've might forgotten that one, tks

**Comment by viniciusdc** [comment]
I just need to add a quick doc about this and we are ready to go :+1: . I've made a simple test with local deployment, this is the structure present on qhub-config:
```yaml
prefect:
  enabled: true
  overrides:
      prefect_agent:
        prefectLabels: [dev, prod]
        job:
          resources:
            requests:
              memory: 20Gi
```

and this one is the output result on k9s:
```yaml
Environment:                                                                                                                                                                                                                                                                                                                                                                                       
│       PREFECT__CLOUD__AGENT__LABELS:         ["dev","prod"]                                                                                                                                                                       │
│       JOB_MEM_REQUEST:                       20Gi                                                                                                                                                                                 

```

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf (hunk 2 above)
Comment (viniciusdc) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/main.tf:4]: This one will be adressed in another PR, see this issue for ref. https://github.com/Quansight/qhub/issues/819

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml [hunk 3]
```diff
@@ -23,6 +23,10 @@ spec:
             - /bin/bash
             - -c
           env:
+            {{- range $k, $v := .Values.envVars }}
+            - name: {{ $k }}
+              value: {{ $v -}}
+            {{ end }}
```
Comment (Adam-D-Lewis) [inline qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:29]: @viniciusdc I added these lines.  They allow you to add arbitrary environment variables to the prefect agent.  This was one thing that was requested by jhamman, and may simplify this PR.  This also allows the users to specify any environment variable including others that could affect prefect without requiring us to make a PR to qhub in the future.  

E.g. A values file like the one below . . . 
```yaml
# values.yaml
prefectImage: <placeholder>
namespace: <placeholder>
serviceAccount: prefect
prefectToken: "<placeholder>"
cloudApi: "<placeholder>"
env: 
  MY_ENV_VAR: my_env_var_value
  MY_ENV_VAR2: my_env_var_value2
  MY_ENV_VAR3: my_env_var_value3
```
. . . would produce the following kubernetes manifest.
```
# Source: prefect/templates/prefect.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: prefect-agent
  name: prefect-agent
spec:
  replicas: 1
  selector:
    matchLabels:
      app: prefect-agent
  template:
    metadata:
      labels:
        app: prefect-agent
    spec:
      serviceAccountName: prefect
      containers:
        - args:
            - prefect agent start kubernetes -e JUPYTERHUB_API_TOKEN=$JUPYTERHUB_API_TOKEN
          command:
            - /bin/bash
            - -c
          env:
            - name: MY_ENV_VAR
              value: my_env_var_value
            - name: MY_ENV_VAR2
              value: my_env_var_value2
            - name: MY_ENV_VAR3
              value: my_env_var_value3
# . . . rest of file would be below
```

File: docs/source/admin_guide/prefect.md [hunk 4]
```diff
@@ -33,6 +33,34 @@ The `TF_VAR_prefect_token` API key is set as `PREFECT__CLOUD__AGENT__AUTH_TOKEN`
...
+- PREFECT__CLOUD__AGENT__LABELS
+- PREFECT__CONTEXT__SECRETS__GCP_CREDENTIALS
+- JOB_MEM_REQUEST
+- JOB_MEM_LIMIT
+- JOB_CPU_REQUEST
+- JOB_CPU_LIMIT
+- IMAGE_PULL_POLICY
+```
+
+For example, if you just want to override the amount of CPU limits for each job, you would need to craft a declarative configuration, in you qhub-config.yaml file, as follows:
```
Comment (Adam-D-Lewis) [inline docs/source/admin_guide/prefect.md:52]: typo: `you` should be `your

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml [hunk 5]
```diff
@@ -38,19 +42,21 @@ spec:
...
+              value: {{ .Values.prefect_agent.job.resources.limits.memory | quote }}
             - name: JOB_CPU_REQUEST
-              value: ''
+              value: {{ .Values.prefect_agent.job.resources.requests.cpu | quote }}
             - name: JOB_CPU_LIMIT
-              value: ''
+              value: {{ .Values.prefect_agent.job.resources.limits.cpu | quote }}
             - name: IMAGE_PULL_POLICY
-              value: ''
+              value: {{ .Values.prefect_agent.job.imagePullPolicy | quote }}
```
Comment (Adam-D-Lewis) [inline qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:59]: Does setting these environment variables actually do anything?  Is this a prefect thing?  If so, can you point me to the documentation describing them please?  Thanks!

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml (hunk 3 above)
Comment (viniciusdc) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:29]: Oh that's nice! thanks for the pointer Adam

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml (hunk 5 above)
Comment (viniciusdc) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:59]: Sure, [this one](https://docs.prefect.io/orchestration/agents/kubernetes.html) exaplains the Agent role in Prefect and [this one](https://docs.prefect.io/orchestration/agents/overview.html#environment-variables) talks about the environment variables usage

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml (hunk 3 above)
Comment (viniciusdc) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:29]: My only concern with that is related with the variables that need use `toJson` when included in the manifest, e.g. `PREFECT__CLOUD__AGENT__LABELS` which is a list and `IMAGE_PULL_SECRETS` which is a dictionary containing the secrets. How would I handle that using the proposed env structure above?

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml (hunk 5 above)
Comment (Adam-D-Lewis) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:59]: Thanks, I can't find for example, JOB_MEM_REQUEST anywhere in their documentation.  How did you find about that env var in particular?

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml (hunk 3 above)
Comment (Adam-D-Lewis) [reply qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:29]: I imagine you could get the desired format by using back slashes.  If not, I think it's reasonable to still keep the environment variables that need special formatting in there.

File: docs/source/admin_guide/prefect.md [hunk 6]
```diff
@@ -33,6 +33,34 @@ The `TF_VAR_prefect_token` API key is set as `PREFECT__CLOUD__AGENT__AUTH_TOKEN`
...
+
+```yaml
+prefect:
+  enabled: true
+  overrides:
+      prefect_agent:
+        job:
+          resources:
+            limit:
+              cpu: 4
```
Comment (Adam-D-Lewis) [inline docs/source/admin_guide/prefect.md:62]: escape characters might help:

  overrides:
      prefect_agent:
        job:
          resources:
            limit:
              cpu: 4
  env:
     MY_ENV_VAR: \[my_value, my_value2\]

**Comment by viniciusdc** [comment]
So based on @Adam-D-Lewis comment and some talks I had with him and @costrouc about the agent helm chart as a whole. I've decided to maintain the overrides values from the `values.yaml` files, using the whole structure block for the variable. But I also used the loop block Adam suggested to permor inclusions of new variables into the agent, this way, for example, in the qhub-config we have now under prefect field:

```yaml
overrides:
      extraVars:
        PREFECT__CONTEXT__SECRETS__AWS_CREDENTIALS: eyJBQ0NFU1NfS0VZIjogImFiY2RlZiIsICJTRUNSRVRfQUNDRVNTX0tFWSI6ICJnaGlqa2xtbiJ9Cg==
      prefect_agent:
        prefectLabels: [dev, prod]
        job:
          resources:
            requests:
              memory: 20Gi
```
where in this case we will override the `PREFECT__CLOUD__AGENT__LABELS` and `JOB_MEM_REQUEST` fields in the chart as well as adding a new variable as an extra, called `PREFECT__CONTEXT__SECRETS__GCP_CREDENTIALS`. We ended up encoding base64 to maintain the string characteristics until reaching the helm chart.

The result can be seen in the cluster deployment for the agent:

```yaml
PREFECT__CLOUD__AGENT__LABELS:               ["dev","prod"] 
JOB_MEM_REQUEST:                             20Gi
PREFECT__CONTEXT__SECRETS__AWS_CREDENTIALS:  {"ACCESS_KEY": "abcdef", "SECRET_ACCESS_KEY": "ghijklmn"}
```

**Comment by Adam-D-Lewis** [comment]
For extraVars, we'd need to add some documentation explaining it.  Also, maybe call it envVars or extraEnvVars to make it clear that these are environment variables?

File: docs/source/admin_guide/prefect.md [hunk 7]
```diff
@@ -33,6 +33,48 @@ The `TF_VAR_prefect_token` API key is set as `PREFECT__CLOUD__AGENT__AUTH_TOKEN`
...
+to authenticate to GCP automatically by adding that specific `GCP_CREDENTIALS` key value pair into your secrets context like so:
+
+```yalm
+prefect:
+  enabled: true
+  overrides:
+    extraEnvVars:
+        PREFECT__CONTEXT__SECRETS__GCP_CREDENTIALS: '<Your value>'
+```
+It's important to keep in mind that all variables defined under `extraEnvVars` must be base64 encoded before passing it's value to the qhub-config file.
```
Comment (Adam-D-Lewis) [inline docs/source/admin_guide/prefect.md:78]: Why do they all need to be base64 encoded?  Is there no way around that?  Seems very inconvenient as a developer.

Comment (viniciusdc) [reply docs/source/admin_guide/prefect.md:78]: I agree with you in that, unfortunately every time we pass the correct structure via cookiecutter to terraform, it understands it as a string, but if the variable has some special characteristics, it is a list or a json, the helm chart will not parse the string correctly resulting in some errors as
```
Error: ---> error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array into Go value of type map[string]interface {} [{'PREFECT__CLOUD__AGENT__LABELS': ['dev', 'prod'], 'JOB_MEM_REQUEST': '20Gi'}]
```

Comment (leej3) [reply docs/source/admin_guide/prefect.md:78]: > a rush to get this PR in (are we?)

Nope. We are rebasing onto this branch so no pressure to merge the PR.

Comment (Adam-D-Lewis) [reply docs/source/admin_guide/prefect.md:78]: I made some progress on this.  I'll push the updates in a commit tomorrow.

**Comment by Adam-D-Lewis** [comment]
I've made some adjustments that don't require base64 encoding of environment variables.  Check it out, @viniciusdc and let me know if this meets all your needs.  I made a repo where I was testing out a local deployment of this branch if that's helpful for you.

https://github.com/Quansight/grafana-prometheus-qhub/tree/prefect_overrides

**Comment by viniciusdc** [comment]
> I've made some adjustments that don't require base64 encoding of environment variables. Check it out, @viniciusdc and let me know if this meets all your needs. I made a repo where I was testing out a local deployment of this branch if that's helpful for you. You can access it at https://github.com/Quansight/grafana-prometheus-qhub/tree/prefect_overrides

Hi @Adam-D-Lewis thank you so much for that! I noticed that if we pass a dic as a key in the overrides envVars, e.g `{  }`, we end up getting a `map[]` object in the final manifest, but this is easily resolved if we pass the original dict as a string. (I will add this to the docs later).

Also, I've included an extra `secretsEnvVars` field in the overrides. This way the user can set a secret that will be added to the manifest chart as a kubernetes secret. Let me know your thoughts about that as well. I've teste locally and it worked pretty well.

File: qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml [hunk 8]
```diff
@@ -18,11 +18,26 @@ spec:
...
             - -c
           env:
+            {{- range $k, $v := .Values.envVars }}
+            - name: {{ $k }}
+              value: {{ squote $v -}}
+            {{ end }}
+            # {{- range $k, $v := .Values.secretEnvVars }}
+            # - name: {{ $k }}
+            #   value: {{ squote $v -}}
+            # {{ end }}
```
Comment (viniciusdc) [inline qhub/template/{{ cookiecutter.repo_directory }}/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml:33]:
```suggestion
```
Remove commented code

**Review by costrouc — APPROVED** [review]

**Pull request merged**